### ISSUE TITLE
Why does my evaluation code yield a parameter count of 39k for EfficientNet-B0, instead of the expected 3.6m?

### ISSUE THREAD

**Issue by penglonghh** [boyagesmile/DNA-Depth#2]
Hello,thank you for your great code.

I evaluate the model b0 by the evaluate.py form Lite-mono, it turns out the params of the whole models is just 594k. The params of the encoder is 39k.But, in your paper ,the number is 4.15M. Why does this situation occur?

**Comment by penglonghh** [comment]
def profile_once(model, x):
    x = x[0, :, :, :].unsqueeze(0)
    flops, params = profile(model, inputs=(x,), verbose=False)
    flops, params = clever_format([flops, params], "%.3f")
    return flops, params
I used this code to calculate the flops and params.

**Comment by boyagesmile** [comment]
4.15M encompasses all the parameters of both the encoder and the decoder, not merely the parameters of the encoder alone.

**Comment by Asuna88** [comment]
我遇到了类似的情况，这是我的测试脚本。

from thop import profile
from thop import clever_format 
           
            # ###############每次保存为模型，测试下这个模型的 GFLOPs，parmas###############tart
            # ##################### 参数量、 GFLOPs 测试############### start
            # # ############ GFLOPs ########### start
            # verbose=False 表示关闭详细日志
       
            input_color = input_image
            print("input_color.shape",input_color.shape)
            flops_e, params_e = profile(encoder, inputs=(input_color, ), verbose=False)
            flops_d, params_d = profile(depth_decoder, inputs=(encoder(input_color), ), verbose=False)

            flops, params = clever_format([flops_e + flops_d, params_e + params_d], "%.3f")
            flops_e, params_e = clever_format([flops_e, params_e],"%.3f")
            flops_d, params_d = clever_format([flops_d, params_d],"%.3f")
            
            # 这里注释掉，避免每次都在打印，在最后打印一次即可
            print("flops_e, params_e") 
            print(flops_e, params_e)    
            print("flops_d, params_d") 
            print(flops_d, params_d)                 
            # 计算total的参数量
            print("total_flops, total_params") 
            print(flops, params) 
            # ############ GFLOPs ########### end
            # ##################### 参数量、 GFLOPs 测试##############end
            

$ bash my_eval_times.sh
-> Loading model from  models/mono_640x192
   Loading pretrained encoder
   Loading pretrained decoder
-> Predicting on 1 test images
input_color.shape torch.Size([1, 3, 192, 640])

flops_e, params_e
65.388M 39.456K

flops_d, params_d
8.914G 554.564K

total_flops, total_params
8.979G 594.020K
   Processed 1 of 1 images - saved predictions to:
   - test_imgs/0000000044_disp.png
   - test_imgs/0000000044_disp.npy
-> Done!

我测试的结果是：

encoder的参数量是39.456K，encoder的计算量是65.388M
decoder的参数量是554.564K，decoder的计算量是8.914G 
encoder + decoder 的参数量是594.020K，encoder + decoder的计算量是8.979G 

我和上面那位同学的疑问一模一样。这是实际代码测试出来的。
感觉您的这个回复好像无法解释这个疑问。

**Comment by Asuna88** [comment]
您回复提到的（encoder+decoder=4.15M，和下面用代码测试的参数量总和594.020K也不吻合）。
（除非您公布的efficientnet.py和depth_decoder.py 有误，或者进行了额外的更新）


我觉得应该按照如下这个数据来写，这个测试的才是准确的
目前测试了N次，结果都是这样的：

encoder的参数量是39.456K，encoder的计算量是65.388M
decoder的参数量是554.564K，decoder的计算量是8.914G
encoder + decoder 的参数量是594.020K，encoder + decoder的计算量是8.979G

注：如果我的思考逻辑、或者测试有误，欢迎大家指出。谢谢大家。

**Comment by boyagesmile** [comment]
问题应该出在efficientnet的encoder上，在[EfficientNet](https://github.com/lukemelas/EfficientNet-PyTorch)中给出的官方测试的参数量为5.3M，我使用的时候去掉了后面的全连接层，所以参数量略有降低，但是不会变成39.456K。
你测试的代码使用的是thop，他对有的组件需要自定义计算量，因为EffciientNet有的层是自定义层，所以thop会跳过计算。我更新了我当时的测试代码 [caculate_flops.py](https://github.com/boyagesmile/DNA-Depth/blob/main/caculate_flops.py)，
输出如下：

> [INFO] Register count_convNd() for <class 'torch.nn.modules.conv.Conv2d'>.
[INFO] Register count_bn() for <class 'torch.nn.modules.batchnorm.BatchNorm2d'>.
EfficientNet encoder: 3.60 M, 0.38G FLOPS
[INFO] Register count_convNd() for <class 'torch.nn.modules.conv.Conv2d'>.
[INFO] Register count_bn() for <class 'torch.nn.modules.batchnorm.BatchNorm2d'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.dropout.Dropout'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.container.Sequential'>.
[INFO] Register count_adap_avgpool() for <class 'torch.nn.modules.pooling.AdaptiveAvgPool2d'>.
[INFO] Register count_linear() for <class 'torch.nn.modules.linear.Linear'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.activation.ReLU'>.
EfficientNet: 4.15 M, 9.88G FLOPS
Flops:  9.88 GMac
Params: 4.15 M
down

**Comment by Asuna88** [comment]
感谢您的解释，应该和测试工具有一定的关系。

不过，我发现个疑问，刚刚下载了您了代码并运行它caculate_flops.py
(https://github.com/boyagesmile/DNA-Depth/blob/main/caculate_flops.py，

发现用这个脚本测试后依旧和您显示的不一致，是哪里的参数信息写错了么？

import torch
import torch.nn as nn
import thop

from efficientnet_pytorch.utils import Conv2dDynamicSamePadding
from efficientnet_pytorch.utils import Conv2dStaticSamePadding
from efficientnet_pytorch.utils import MemoryEfficientSwish
from thop.vision.basic_hooks import count_convNd, zero_ops

import networks
from ptflops.flops_counter import get_model_complexity_info

from torchstat import stat

class model_name(nn.Module):
    def __init__(self, encoder, decoder):
        super(model_name, self).__init__()
        self.encoder = encoder
        self.depth_decoder = decoder
    def forward(self, x):
        feat = self.encoder(x)
        result = self.depth_decoder(feat)

        return(result)

def count_sigmoid(m, x, y):
    """
    Using this approximation for exponetial operation:  exp(x) = 1 + x + x^2/2! + .. + x^9/9!
    For sigmoid f(x) = 1/(1+exp(x)): there are totally 10 add ops, 9 division ops(2! are considered as constant).
    Since it is element-wise operation. The final ops is about(10+9)*num_elements.
    """
    x = x[0]

    nelements = x.numel()

    total_ops = 19 * nelements
    m.total_ops += torch.Tensor([int(total_ops)])

if __name__ == '__main__':
    custom_ops = {
            Conv2dDynamicSamePadding: count_convNd,
            Conv2dStaticSamePadding: count_convNd,
            MemoryEfficientSwish: zero_ops,
        }

    encoder = networks.EfficientEncoder()
    depth_decoder = networks.EfficientDecoder(encoder.num_ch_enc)
    model = model_name(encoder, depth_decoder)
    verbose = False
    x = torch.randn(1, 3, 192, 640)
    args = dict(inputs=(x,), custom_ops=custom_ops, verbose=True)
    macs, params = thop.profile(encoder, **args)
    print(f'EfficientNet encoder: {params * 1e-6:.2f} M, {macs * 1e-9:.2f}G FLOPS')

    x = torch.randn(1, 3, 192, 640)
    args = dict(inputs=(x,), custom_ops=custom_ops, verbose=True)
    macs, params = thop.profile(model, **args)
    print(f'EfficientNet: {params * 1e-6:.2f} M, {macs * 1e-9:.2f}G FLOPS')

    # flops, params = get_model_complexity_info(model, (3, 192, 640), as_strings=True, print_per_layer_stat=True)
    flops, params = get_model_complexity_info(model, (3, 192, 640), as_strings=True, print_per_layer_stat=False)
    print('Flops:  ' + flops)
    print('Params: ' + params)
    print('down')



我的测试结果：
$ python caculate_flops.py

[INFO] Register zero_ops() for <class 'torch.nn.modules.padding.ZeroPad2d'>.
[INFO] Register count_normalization() for <class 'torch.nn.modules.batchnorm.BatchNorm2d'>.
EfficientNet encoder: 0.04 M, 0.07G FLOPS
[INFO] Register zero_ops() for <class 'torch.nn.modules.padding.ZeroPad2d'>.
[INFO] Register count_normalization() for <class 'torch.nn.modules.batchnorm.BatchNorm2d'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.dropout.Dropout'>.
[INFO] Register count_convNd() for <class 'torch.nn.modules.conv.Conv2d'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.container.Sequential'>.
[INFO] Register count_adap_avgpool() for <class 'torch.nn.modules.pooling.AdaptiveAvgPool2d'>.
[INFO] Register count_linear() for <class 'torch.nn.modules.linear.Linear'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.activation.ReLU'>.
EfficientNet: 0.59 M, 8.98G FLOPS
Flops:  9.02 GMac
Params: 4.15 M
down


我的测试环境
Python 3.10.18 (main, Jun  5 2025, 13:14:17) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'2.8.0+cu128'
>>> 

为什么还是和您展示的不一致？

**Comment by boyagesmile** [comment]
我用的torch版本是1.7.0，显卡是RTX3090。我尝试运行了你的代码，输出是：

> [INFO] Register count_convNd() for <class 'torch.nn.modules.conv.Conv2d'>.
[INFO] Register count_bn() for <class 'torch.nn.modules.batchnorm.BatchNorm2d'>.
EfficientNet encoder: 3.60 M, 0.92G FLOPS
[INFO] Register count_convNd() for <class 'torch.nn.modules.conv.Conv2d'>.
[INFO] Register count_bn() for <class 'torch.nn.modules.batchnorm.BatchNorm2d'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.dropout.Dropout'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.container.Sequential'>.
[INFO] Register count_adap_avgpool() for <class 'torch.nn.modules.pooling.AdaptiveAvgPool2d'>.
[INFO] Register count_linear() for <class 'torch.nn.modules.linear.Linear'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.activation.ReLU'>.
EfficientNet: 4.15 M, 9.88G FLOPS
Flops:  9.88 GMac
Params: 4.15 M
down

我不清楚你的运行环境，我这里两种测试和方法的输出是一致的。
另外，你的代码给我的感觉是，thop中的自定义变量没有设置成功，`[INFO] Register count_convNd() for <class 'torch.nn.modules.conv.Conv2d'>`.没有成功注册，因此与ptflops输出的差距偏大。比起torch版本，或许你更应该关注thop的版本，我用的版本是0.0.5-2205040957，ptflops的版本是 0.4.1，希望这些能对你有所帮助。

**Comment by Asuna88** [comment]
好的，谢谢您的解释。这是我目前的安装包版本
 $ pip list | grep thop
thop                          0.1.1.post2209072238
 $ pip list | grep ptflops
ptflops                       0.7.5

空些时候，我尝试按照你的版本重装后，测试下再来反馈。

感谢您精彩的工作和代码！
谢谢

**Comment by Asuna88** [comment]
哦，对了，确认下。

在这份代码参数设置下，（图像输入尺寸为 192*640）

x = torch.randn(1, 3, 192, 640)
args = dict(inputs=(x,), custom_ops=custom_ops, verbose=True)
macs, params = thop.profile(encoder, **args)
print(f'EfficientNet encoder: {params * 1e-6:.2f} M, {macs * 1e-9:.2f}G FLOPS')

x = torch.randn(1, 3, 192, 640)
args = dict(inputs=(x,), custom_ops=custom_ops, verbose=True)
macs, params = thop.profile(model, **args)
print(f'EfficientNet: {params * 1e-6:.2f} M, {macs * 1e-9:.2f}G FLOPS')

您这边计算的结果是（对于图像输入尺寸为 192*640而言）：

encoder 参数量：3.60 M； encoder 计算量 0.92G FLOPS
总体模型的参数量：4.15 M； 总体模型的计算量 9.88G FLOPS

因此可以推断：
decoder的参数量：0.55 M； encoder 计算量 8.96G FLOPS

您看下， 没问题吧。

**Comment by boyagesmile** [comment]
差不多是这样的，单独计算解码器也是可以的。解码器没有加自定义模块。

**Comment by Asuna88** [comment]
感谢，我还没来得及复现下，安装好软件复现后我再来汇报
